src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -160,39 +196,53 @@ def _get_clone_source(self):
 
     @property
     def status(self):
-        state = self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE)
-        subvolume_type = self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_TYPE)
+        state = SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
+        subvolume_type = self.type
         subvolume_status = {
-            'state' : state
+            'state' : state.value
         }
-        if not OpSm.is_final_state(state) and subvolume_type == SubvolumeBase.SUBVOLUME_TYPE_CLONE:
+        if not SubvolumeOpSm.is_complete_state(state) and subvolume_type == SubvolumeTypes.TYPE_CLONE:
             subvolume_status["source"] = self._get_clone_source()
         return subvolume_status
 
     @property
     def state(self):
-        return self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE)
+        return SubvolumeStates.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_STATE))
 
     @state.setter
     def state(self, val):
-        state = val[0]
+        state = val[0].value
         flush = val[1]
         self.metadata_mgr.update_global_section(MetadataManager.GLOBAL_META_KEY_STATE, state)
         if flush:
             self.metadata_mgr.flush()
 
-    def remove(self):
+    @property
+    def type(self):
+        return SubvolumeTypes.from_value(self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_TYPE))
+
+    def remove(self, retainsnaps=False):
+        if retainsnaps:
+            raise VolumeException(-errno.EINVAL, "subvolume '{0}' does not support snapshot retention on delete".format(self.subvolname))
+        if self.list_snapshots():
+            raise VolumeException(-errno.ENOTEMPTY, "subvolume '{0}' has snapshots".format(self.subvolname))
         self.trash_base_dir()
 
     def resize(self, newsize, noshrink):
         subvol_path = self.path
         return self._resize(subvol_path, newsize, noshrink)
 
+    def info(self):
```

Comment:
Should the pattern be to define `info` in the base class, and not implement the method in v1, and use it in v2 appropriately calling it via `super(SubvolumeBase, self).info`?

Reply:
Updated with the super pattern.